### ISSUE TITLE
Duplicate builds in make -j

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#451]
This is a followup to #400. I noticed that 'make -j' in the .mad directories triggers a double build of some targets. This may also eventually result in build errors. There are some dependencies that need to be fixed.

Example in ggtt.mad
```
...
ar cru  ../../lib/libcernlib.a  abend.o dlsqp2.o lenocc.o mtlprt.o mtlset.o radmul.o
ranlib  ../../lib/libcernlib.a
make[2]: Leaving directory `/data/avalassi/GPU2020/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/Source/CERNLIB'
ar cru  ../../lib/libmodel.a  couplings.o lha_read.o printout.o rw_para.o model_functions.o couplings1.o couplings2.o
ranlib  ../../lib/libmodel.a
make[2]: Leaving directory `/data/avalassi/GPU2020/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/Source/MODEL'
ar cru  ../../lib/libdhelas.a  aloha_functions.o FFV1_1.o FFV1_0.o FFV1_2.o VVV1P0_1.o
ranlib  ../../lib/libdhelas.a
make[2]: Leaving directory `/data/avalassi/GPU2020/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/Source/DHELAS'
make[1]: Entering directory `/data/avalassi/GPU2020/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx'
make -C ../../src  -f Makefile
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gcheck_sa.cu -o gcheck_sa.o
ar cru  ../../lib/libpdf.a  Ctq6Pdf.o pdfwrap.o opendata.o pdf.o PhotonFlux.o pdg2pdf.o NNPDFDriver.o eepdf.o gridpdfaux.o dfint.o kerset.o ElectroweakFluxDriver.o ElectroweakFlux.o
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gCPPProcess.cu -o gCPPProcess.o
ranlib  ../../lib/libpdf.a
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gMatrixElementKernels.cu -o gMatrixElementKernels.o
make[2]: Leaving directory `/data/avalassi/GPU2020/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/Source/PDF'
make[1]: Entering directory `/data/avalassi/GPU2020/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx'
make -C ../../src  -f Makefile
AVX=512y
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gcheck_sa.cu -o gcheck_sa.o
AVX=512y
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gBridgeKernels.cu -o gBridgeKernels.o
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gCPPProcess.cu -o gCPPProcess.o
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fno-fast-math -Xcompiler -fPIC -c gCrossSectionKernels.cu -o gCrossSectionKernels.o
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c gMatrixElementKernels.cu -o gMatrixElementKernels.o
make[2]: Entering directory `/data/avalassi/GPU2020/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/src'
ccache /usr/local/cuda-11.6/bin/nvcc  -O3  -lineinfo -I. -I../../src -I../../../../../tools -I/usr/local/cuda-11.6/include/ -DUSE_NVTX -gencode arch=compute_70,code=compute_70 -gencode arch=compute_70,code=sm_70 -use_fast_math -std=c++17  -ccbin /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -DMGONGPU_FPTYPE_DOUBLE -Xcompiler -fPIC -c -x cu fbridge.cc -o fbridge_cu.o
...
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -o check.exe ./check_sa.o  -ldl -pthread -L../../lib -lmg5amc_common -Wl,-rpath,'$ORIGIN/../../lib'  -L/usr/local/cuda-11.6/lib64/ -lcurand  -L../../lib -lmg5amc_gg_ttx_cpp ./RandomNumberKernels.o ./RamboSamplingKernels.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -o fcheck.exe ./fcheck_sa.o ./fsampler.o -L../../lib -lmg5amc_common -Wl,-rpath,'$ORIGIN/../../lib'  -L/usr/local/cuda-11.6/lib64/ -lcurand  -lgfortran -L../../lib -lmg5amc_gg_ttx_cpp ./RandomNumberKernels.o ./RamboSamplingKernels.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -o check.exe ./check_sa.o  -ldl -pthread -L../../lib -lmg5amc_common -Wl,-rpath,'$ORIGIN/../../lib'  -L/usr/local/cuda-11.6/lib64/ -lcurand  -L../../lib -lmg5amc_gg_ttx_cpp ./RandomNumberKernels.o ./RamboSamplingKernels.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -o fcheck.exe ./fcheck_sa.o ./fsampler.o -L../../lib -lmg5amc_common -Wl,-rpath,'$ORIGIN/../../lib'  -L/usr/local/cuda-11.6/lib64/ -lcurand  -lgfortran -L../../lib -lmg5amc_gg_ttx_cpp ./RandomNumberKernels.o ./RamboSamplingKernels.o
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.34-990b2/x86_64-centos7/bin/ld: ./check_sa.o: in function `main.cold':
check_sa.cc:(.text.unlikely+0x456): undefined reference to `VTT for mg5amcCpu::CrossSectionKernelHost'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.34-990b2/x86_64-centos7/bin/ld: check_sa.cc:(.text.unlikely+0x639): undefined reference to `mg5amcCpu::CPPProcess::~CPPProcess()'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.34-990b2/x86_64-centos7/bin/ld: ./check_sa.o: in function `main':
check_sa.cc:(.text.startup+0x3a5): undefined reference to `mg5amcCpu::MatrixElementKernelHost::hostSupportsSIMD(bool)'
/cvmfs/sft.cern.ch/lcg/releases/binutils/2.34-990b2/x86_64-centos7/bin/ld: check_sa.cc:(.text.startup+0x4dd): undefined reference to `mg5amcCpu::CPPProcess::CPPProcess(bool, bool)'
...
```

**Comment by valassi** [comment]
This is a nice place where Makefile grouped targets would be useful.
Unofrtunately this is only a make 4.3 feature https://stackoverflow.com/questions/61345698/gnu-make-grouped-targets-are-not-grouped
And I am currently using make 3.8...

**Comment by valassi** [comment]
One alternative I was thinking of is to have a dummy file target that just signals a full directory is done (typically, make Source). This is essentially the solution proposed here
https://stackoverflow.com/questions/63928066/what-is-the-canonical-way-to-simulate-grouped-targets-in-older-versions-of-make

**Comment by valassi** [comment]
I have committed a fix in 3071ec4

**Comment by valassi** [comment]
I now have another issue I want to fix in that same PR #452 
```
cd ../../Source; make
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c myamp.f -I../../Source/ -fopenmp -o myamp.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c genps.f -I../../Source/ -fopenmp -o genps.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c unwgt.f -I../../Source/ -fopenmp -o unwgt.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c setcuts.f -I../../Source/ -fopenmp -o setcuts.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c get_color.f -I../../Source/ -fopenmp -o get_color.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c cuts.f -I../../Source/ -fopenmp -o cuts.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c cluster.f -I../../Source/ -fopenmp -o cluster.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c reweight.f -I../../Source/ -fopenmp -o reweight.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c initcluster.f -I../../Source/ -fopenmp -o initcluster.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c addmothers.f -I../../Source/ -fopenmp -o addmothers.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c setscales.f -I../../Source/ -fopenmp -o setscales.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c idenparts.f -I../../Source/ -fopenmp -o idenparts.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c dummy_fct.f -I../../Source/ -fopenmp -o dummy_fct.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c driver.f -I../../Source/ -fopenmp -o driver.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c auto_dsig.f -I../../Source/ -fopenmp -o auto_dsig.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp  -c auto_dsig1.f -I../../Source/ -fopenmp -o auto_dsig1.o
make[1]: Entering directory `/data/avalassi/GPU2020/madgraph4gpuX/epochX/cudacpp/ee_mumu.mad/Source'
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132  -c -o StringCast.o StringCast.f
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132  -c -o ranmar.o ranmar.f
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132 -w -cpp   -c matrix1.f -I../../Source/ -fopenmp
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132  -c -o alfas_functions.o alfas_functions.f
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/gfortran -w -fPIC  -ffixed-line-length-132  -c -o transpole.o transpole.f
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/10.2.0-c44b3/x86_64-centos7/bin/g++ -std=c++11 -Wall -Wshadow -Wextra -c counters.cpp -o counters.o
genps.f:92:10:

   92 |       use DiscreteSampler
      |          1
Fatal Error: Cannot open module file ‘discretesampler.mod’ for reading at (1): No such file or directory
compilation terminated.
```

**Comment by valassi** [comment]
The issue above was not reproducible also because .mod were not being removed in make clean, will fix that too

**Comment by valassi** [comment]
This is now fixed by PR #452 - closing

**Comment by valassi** [comment]
I have had more such issues with duplicate builds (and crashes) after adding 'make avxall -j' as part of PR #454 

About 'make -j', note the explanation by the current maintainer (IIUC) of gnu make about how this works
https://stackoverflow.com/a/47459209

Note also I am currently using a very old make 3.81. Many features that could be useful only exist in newer versions. For instance grouped targets are only in make 4.3: https://stackoverflow.com/a/61345989
